rust/src/tokio/read.rs

```diff
@@ -0,0 +1,380 @@
+            }
+            if self.footer_seen && !self.options.skip_end_magic {
+                reader.read_exact(&mut self.scratch[..MAGIC.len()]).await?;
+                if &self.scratch[..MAGIC.len()] != MAGIC {
+                    return Err(McapError::BadMagic);
+                }
+                return Ok(Cmd::Stop);
+            }
+            let readlen = reader.read(&mut self.scratch[..9]).await?;
+            if readlen == 0 && self.options.skip_end_magic {
+                return Ok(Cmd::Stop);
+            }
+            if readlen != 9 {
+                return Err(McapError::UnexpectedEof);
+            }
+            let opcode = self.scratch[0];
+            if opcode == records::op::FOOTER {
+                self.footer_seen = true;
+            }
+            let record_len = byteorder::LittleEndian::read_u64(&self.scratch[1..9]);
+            if opcode == records::op::CHUNK && !self.options.emit_chunks {
+                let header = read_chunk_header(reader, data, record_len).await?;
+                return Ok(Cmd::EnterChunk {
+                    header,
+                    len: record_len,
+                });
+            }
+            data.resize(record_len as usize, 0);
+            reader.read_exact(&mut data[..]).await?;
+            Ok(Cmd::YieldRecord(opcode))
+        } else {
+            let len = self.reader.read(&mut self.scratch[..9]).await?;
+            if len == 0 {
+                return Ok(Cmd::ExitChunk);
+            }
+            if len != 9 {
+                return Err(McapError::UnexpectedEof);
+            }
+            let opcode = self.scratch[0];
+            let record_len = byteorder::LittleEndian::read_u64(&self.scratch[1..9]);
+            data.resize(record_len as usize, 0);
+            self.reader.read_exact(&mut data[..]).await?;
+            Ok(Cmd::YieldRecord(opcode))
+        }
+    }
+}
+
+async fn read_chunk_header<R: AsyncRead + std::marker::Unpin>(
+    reader: &mut R,
+    scratch: &mut Vec<u8>,
+    record_len: u64,
+) -> McapResult<records::ChunkHeader> {
+    let mut header = records::ChunkHeader {
+        message_start_time: 0,
+        message_end_time: 0,
+        uncompressed_size: 0,
+        uncompressed_crc: 0,
+        compression: String::new(),
+        compressed_size: 0,
+    };
+    if record_len < 40 {
+        return Err(McapError::RecordTooShort {
+            opcode: records::op::CHUNK,
+            len: record_len,
+            expected: 40,
+        });
+    }
+    scratch.resize(32, 0);
+    reader.read_exact(&mut scratch[..]).await?;
+    header.message_start_time = byteorder::LittleEndian::read_u64(&scratch[0..8]);
+    header.message_end_time = byteorder::LittleEndian::read_u64(&scratch[8..16]);
+    header.uncompressed_size = byteorder::LittleEndian::read_u64(&scratch[16..24]);
+    header.uncompressed_crc = byteorder::LittleEndian::read_u32(&scratch[24..28]);
+    let compression_len = byteorder::LittleEndian::read_u32(&scratch[28..32]);
```

Comment:
The `binrw` crate and `BinReaderExt` trait seem to have some convenience functions to make this kind of thing easier:

```rust
    reader.read_exact(&mut scratch[..]).await?;
    
    let mut cursor = std::io::Cursor::new(&scratch);
    
    header.message_start_time = cursor.read_le()?;
    header.message_end_time = cursor.read_le()?;
```

There are similar methods on the `AsyncReadExt` traits and `byteorder::ReadBytesExt` but it's all probably much of a muchness.